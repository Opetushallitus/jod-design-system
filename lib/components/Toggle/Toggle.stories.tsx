import type { StoryObj } from '@storybook/react-vite';
import React from 'react';
import { fn } from 'storybook/test';
import type { TitledMeta } from '../../utils';
import { Toggle } from './Toggle';

const meta = {
  title: 'Forms/Toggle',
  component: Toggle,
  tags: ['autodocs'],
} satisfies TitledMeta<typeof Toggle>;

export default meta;

type Story = StoryObj<typeof meta>;

export const Default: Story = {
  render: (args) => {
    const [checked, setChecked] = React.useState(args.checked);
    React.useEffect(() => {
      setChecked(args.checked);
    }, [args.checked]);
    return <Toggle {...args} checked={checked} onChange={() => setChecked(!checked)} />;
  },
  parameters: {
    design: {
      type: 'figma',
      url: 'https://www.figma.com/design/a3heEFNpoI9MRcMFtbYVfO/Osaamispolku---Design-system?node-id=459-2024&t=ZhYQBTprMjaC7Bhl-0',
    },
    docs: {
      description: {
        story: 'Toggle component for toggling between two states.',
      },
    },
  },
  argTypes: {
    variant: {
      options: ['YKSILO', 'OHJAAJA', 'TIETOPALVELU'],
      control: {
        type: 'radio',
      },
    },
  },
  args: {
    onChange: fn(),
    checked: true,
    ariaLabel: 'Toggle',
    variant: 'yksilo',
    disabled: false,
  },
};

export const Disabled: Story = {
  parameters: {
    design: {
      type: 'figma',
      url: 'https://www.figma.com/design/a3heEFNpoI9MRcMFtbYVfO/Osaamispolku---Design-system?node-id=459-2024&t=ZhYQBTprMjaC7Bhl-0',
    },
    docs: {
      description: {
        story: 'Toggle component in disabled state.',
      },
    },
  },
  args: {
    onChange: fn(),
    disabled: true,
    checked: true,
    ariaLabel: 'Toggle',
    variant: 'yksilo',
  },
};
