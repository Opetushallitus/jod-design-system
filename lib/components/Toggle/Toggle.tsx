import { cx } from '../../cva';
import { Variant } from '../../utils';

export interface ToggleProps {
  ariaLabel: string;
  checked?: boolean;
  disabled?: boolean;
  variant: Variant;
  onChange: (newValue: boolean) => void;
}
export const Toggle = ({ onChange, checked, disabled, ariaLabel, variant }: ToggleProps) => {
  return (
    <button
      onClick={() => !disabled && onChange(!checked)}
      disabled={disabled}
      aria-label={ariaLabel}
      aria-pressed={checked}
      className={cx('ds:transition-all ds:duration-200 ds:w-[52px] ds:h-[32px] ds:relative ds:rounded-2xl ds:flex', {
        'ds:cursor-pointer': !disabled,
        'ds:bg-inactive-gray': !checked || disabled,
        'ds:bg-accent': checked && !disabled && variant === 'YKSILO',
        'ds:bg-secondary-2-dark': checked && !disabled && variant === 'OHJAAJA',
        'ds:bg-secondary-4-dark': checked && !disabled && variant === 'TIETOPALVELU',
      })}
    >
      <span
        className={cx(
          'ds:size-6 ds:bg-white ds:rounded-full ds:m-2 ds:transition-all ds:duration-200 ds:transform ds:translate-x-0',
          { 'ds:translate-x-[20px]': checked },
        )}
      />
    </button>
  );
};
