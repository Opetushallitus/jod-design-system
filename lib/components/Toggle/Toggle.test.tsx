import { fireEvent, render, screen } from '@testing-library/react';
import { describe, expect, it, vi } from 'vitest';
import { Toggle } from './Toggle';

describe('Toggle', () => {
  it('renders as a button', () => {
    const { container } = render(
      <Toggle checked={false} onChange={vi.fn()} ariaLabel="Toggle something" serviceVariant="yksilo" />,
    );
    const button = screen.getByRole('button', { name: /toggle something/i });
    expect(button).toBeInTheDocument();
    expect(container.firstChild).toMatchSnapshot();
  });

  it('shows correct aria-pressed state', () => {
    const { rerender } = render(
      <Toggle checked={false} onChange={vi.fn()} ariaLabel="Toggle" serviceVariant="yksilo" />,
    );
    const button = screen.getByRole('button');
    expect(button).toHaveAttribute('aria-pressed', 'false');

    rerender(<Toggle checked={true} onChange={vi.fn()} ariaLabel="Toggle" serviceVariant="yksilo" />);
    expect(button).toHaveAttribute('aria-pressed', 'true');
  });

  it('calls onChange with toggled value when clicked', () => {
    const handleChange = vi.fn();
    render(<Toggle checked={false} onChange={handleChange} ariaLabel="Toggle" serviceVariant="yksilo" />);
    const button = screen.getByRole('button');
    fireEvent.click(button);
    expect(handleChange).toHaveBeenCalledWith(true);
  });

  it('does not call onChange when disabled', () => {
    const handleChange = vi.fn();
    render(<Toggle checked={false} onChange={handleChange} disabled ariaLabel="Toggle" serviceVariant="yksilo" />);
    const button = screen.getByRole('button');
    fireEvent.click(button);
    expect(handleChange).not.toHaveBeenCalled();
  });

  it('is disabled when the disabled prop is true', () => {
    render(<Toggle checked={false} onChange={vi.fn()} disabled ariaLabel="Toggle" serviceVariant="yksilo" />);
    const button = screen.getByRole('button');
    expect(button).toBeDisabled();
  });

  it('has the correct aria-label', () => {
    render(<Toggle checked={false} onChange={vi.fn()} ariaLabel="Custom label" serviceVariant="yksilo" />);
    const button = screen.getByRole('button');
    expect(button).toHaveAttribute('aria-label', 'Custom label');
  });

  it('emits data-testid when dataTestId is provided', () => {
    render(
      <Toggle checked={false} onChange={vi.fn()} ariaLabel="Toggle id" serviceVariant="yksilo" dataTestId="tgl" />,
    );
    expect(screen.getByTestId('tgl')).toBeInTheDocument();
  });
});
