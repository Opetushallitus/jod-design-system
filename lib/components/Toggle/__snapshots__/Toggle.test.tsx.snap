// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Toggle > renders as a button 1`] = `
<button
  aria-label="Toggle something"
  aria-pressed="false"
  class="ds:transition-all ds:duration-200 ds:w-[52px] ds:h-[32px] ds:relative ds:rounded-2xl ds:flex ds:cursor-pointer ds:bg-inactive-gray"
>
  <span
    class="ds:size-6 ds:bg-white ds:rounded-full ds:m-2 ds:transition-all ds:duration-200 ds:transform ds:translate-x-0"
  />
</button>
`;
